packages/sdk/test/unit/Mapping.test.ts

```diff
@@ -97,25 +97,35 @@ describe('Mapping', () => {
     })
 
     it('max size', async () => {
-        const SIZE = 3
+        const maxSize = 3
         const valueFactory = jest.fn().mockImplementation(async ([p1, p2]: [string, number]) => {
             return `${p1}${p2}`
         })
-        const mapping = createCacheMap({ valueFactory, maxSize: 3 })
-        const ids = range(SIZE)
+        const mapping = createCacheMap({ valueFactory, maxSize })
+        const ids = range(maxSize)
+        /**
+         * Each call to `get` is considered usage. The following populates the cache with
+         * entries for: foo0, foo1, foo2 (3 in total).
+         */
         for (const id of ids) {
             await mapping.get(['foo', id])
         }
-        expect(valueFactory).toHaveBeenCalledTimes(3)
-        // add a value which is not in cache
+        expect(valueFactory).toHaveBeenCalledTimes(maxSize)
+        /**
+         * Calling `get` on a key that's not in the cache causes usage. It will discard
+         * an item associated to `foo0` key (least recently used).
+         */
         await mapping.get(['foo', -1])
-        expect(valueFactory).toHaveBeenCalledTimes(4)
-        // one of the items was removed from cache when -1 was added, now we is re-add that
-        // (we don't know which item it was)
+        expect(valueFactory).toHaveBeenCalledTimes(maxSize + 1)
+        /**
+         * The current list of keys (most-to-least recently used) goes as follows: foo-1,
+         * foo2, foo1. Going through all ids now will reconstruct the initial collection
+         * causing 3 hits.
+         */
         for (const id of ids) {
             await mapping.get(['foo', id])
         }
-        expect(valueFactory).toHaveBeenCalledTimes(5)
+        expect(valueFactory).toHaveBeenCalledTimes(maxSize + 1 + 3)
```

Comment:
Why this is now 7 instead of 5? Assumed it should be the initial 3 evaluations + evaluation of `foo-1` + re-evaluation of dropped `foo0` 

Also: could use `MAX_SIZE` here in place of 3?

Reply:
This tests does the following
| | Operation | Slot 0 | Slot 1 | Slot 2 |
| - | --------- | ------ | ------ | ------ |
| 1 | `get('foo',  0)` | foo0|||
| 2 | `get('foo',  1)` | foo1|foo0|
| 3 | `get('foo',  2)` | foo2|foo1|foo0|
| 4 | `get('foo', -1)` | foo-1|foo2|foo1|
| 5 | `get('foo',  0)` | foo0|foo-1|foo2|
| 6 | `get('foo',  1)` | foo1|foo0|foo-1|
| 7 | `get('foo',  1)` | foo2|foo1|foo0|

Each row is a hit to the value factory because we happen to always hit an entry that's either not yet in the collection or has just been discarded.

Reply:
Thanks, great explanation 👍